src/Share/Codebase.hs

```diff
@@ -174,14 +186,37 @@ codebaseEnv !_authZReceipt codebaseLoc = do
 
 -- | Construct a Runtime linked to a specific codebase.
 -- Don't use the runtime for one codebase with another codebase.
-codebaseRuntime :: (MonadReader (Env.Env x) m, MonadUnliftIO m) => CodebaseEnv -> m CodebaseRuntime
-codebaseRuntime CodebaseEnv {codebaseOwner} = do
+codebaseRuntime :: (MonadReader (Env.Env x) m, MonadUnliftIO m) => CodebaseEnv -> m (CodebaseRuntime IO)
+codebaseRuntime codebase = do
   unisonRuntime <- asks Env.sandboxedRuntime
-  codeLookup <- codeLookupForUser codebaseOwner
-  toIO <- UnliftIO.askRunInIO
-  let codebaseRt = CodebaseRuntime {codeLookup, cachedEvalResult, unisonRuntime}
-      cachedEvalResult r = (fmap . fmap) Term.unannotate . toIO . PG.runTransaction . loadCachedEvalResult codebaseOwner $ r
-  pure codebaseRt
+  rt <- liftIO (codebaseRuntime' unisonRuntime codebase)
+  unlift <- badAskUnliftCodebaseRuntime
+  pure (unlift rt)
+
+-- | Ideally, we'd use this – a runtime with lookup actions in transaction, not IO. But that will require refactoring to
+-- the runtime interface in ucm, so we can't use it for now. That's bad: we end up unsafely running separate
+-- transactions for inner calls to 'codeLookup' / 'cachedEvalResult', which can lead to deadlock due to a starved
+-- connection pool.
+codebaseRuntime' :: Runtime Symbol -> CodebaseEnv -> IO (CodebaseRuntime (PG.Transaction e))
+codebaseRuntime' unisonRuntime CodebaseEnv {codebaseOwner} = do
+  cacheVar <- newTVarIO (CodeLookupCache mempty mempty)
+  pure
+    CodebaseRuntime
+      { codeLookup = codeLookupForUser cacheVar codebaseOwner,
+        cachedEvalResult = (fmap . fmap) Term.unannotate . loadCachedEvalResult codebaseOwner,
+        unisonRuntime
+      }
+
+badAskUnliftCodebaseRuntime ::
```

Comment:
Can you add a note, either an altered version of the note from `codebaseRuntime'` or something, just to explain why this one is "bad" :)